### DISCUSSION TITLE
Problems with Mailgun SMTP configuration

### DISCUSSION THREAD

**Discussion by gatodelcaribe** [plausible/analytics#3335] [Self-Hosted Support]
I'm using the next text into 'plausible-conf.env' but email isn't working:

MAILER_EMAIL= email
SMTP_HOST_ADDR=smtp.mailgun.org
SMTP_HOST_PORT=465
SMTP_USER_NAME= username email
SMTP_USER_PWD= my password
SMTP_HOST_SSL_ENABLED=false
SMTP_RETRIES 5

**Comment by dimejo** [comment]
Might be related to https://github.com/plausible/analytics/discussions/3299#discussioncomment-6833233. Please try setting `SMTP_HOST_SSL_ENABLED=true`.

**Reply by gatodelcaribe** [reply]
The same, not working.

**Reply by dimejo** [reply]
No logs?

**Reply by gatodelcaribe** [reply]
Can you tell me how to get the logs?

**Reply by dimejo** [reply]
`docker compose logs plausible`
Replace `plausible` with the name of the Plausible service in your `docker-compose.yml` in case you changed it.

**Reply by gatodelcaribe** [reply]
into /opt/plausible when I execute the command I get this on repeat: 

`plausible_1  | Running migrations for Elixir.Plausible.Repo
plausible_1  | Running migrations for Elixir.Plausible.IngestRepo
plausible_1  | Migrations successful!
plausible_1  | Loading plausible..
plausible_1  | Starting dependencies..
plausible_1  | Starting repos..
plausible_1  | create Plausible.Repo database if it doesn't exist
plausible_1  | create Plausible.IngestRepo database if it doesn't exist
plausible_1  | Creation of Db successful!
plausible_1  | Loading plausible..
plausible_1  | Starting dependencies..
plausible_1  | Starting repos..
plausible_1  | Running migrations for Elixir.Plausible.Repo
plausible_1  | Running migrations for Elixir.Plausible.IngestRepo
plausible_1  | Migrations successful!
root@Plausible-analitycs:/opt/plausible# `

My install is done step by step on this tutorial:
https://www.digitalocean.com/community/tutorials/how-to-install-plausible-analytics-on-ubuntu-22-04

And my config.env is into /opt/plausible 

I modify, exit, save, return. And Reboot the droplet.

**Reply by ruslandoga** [reply]
👋 @gatodelcaribe 

Please also note that `MAILER_EMAIL` is the email that you are sending from, not to, and that if you change the env vars, you need to recreate the container, not just restart it, and that the email errors are sent to Sentry and not logged (I'll open a PR to fix that).

https://github.com/plausible/analytics/blob/e30ba8b27f869fcc4063e914da34ec5ae412ab98/lib/plausible/mailer.ex#L13-L31

**Reply by ruslandoga** [reply]
PR: https://github.com/plausible/analytics/pull/3336

**Reply by gatodelcaribe** [reply]
@ruslandoga 

Ok, now with recreating the container I put the line "DISABLE_REGISTRATION=false" and I'm, allowed to register users, so the container so the .env file is working now. 

Trying to send an invitation I get this logs: 

```elixir
plausible_1  | Migrations successful!
plausible_1  | 08:19:22.673 [warning] Description: 'Server authenticity is not verified since certificate path validation is not enabled'
plausible_1  |      Reason: 'The option {verify, verify_peer} and one of the options \'cacertfile\' or \'cacerts\' are required to enable this.'
plausible_1  | 
plausible_1  | 08:19:22.783 [warning] Description: 'Server authenticity is not verified since certificate path validation is not enabled'
plausible_1  |      Reason: 'The option {verify, verify_peer} and one of the options \'cacertfile\' or \'cacerts\' are required to enable this.'
plausible_1  | 
plausible_1  | 08:19:22.889 [warning] Description: 'Server authenticity is not verified since certificate path validation is not enabled'
plausible_1  |      Reason: 'The option {verify, verify_peer} and one of the options \'cacertfile\' or \'cacerts\' are required to enable this.'
plausible_1  | 
plausible_1  | 08:19:22.996 [warning] Description: 'Server authenticity is not verified since certificate path validation is not enabled'
plausible_1  |      Reason: 'The option {verify, verify_peer} and one of the options \'cacertfile\' or \'cacerts\' are required to enable this.'
plausible_1  | 
plausible_1  | 08:19:23.101 [warning] Description: 'Server authenticity is not verified since certificate path validation is not enabled'
plausible_1  |      Reason: 'The option {verify, verify_peer} and one of the options \'cacertfile\' or \'cacerts\' are required to enable this.'
plausible_1  | 
plausible_1  | 08:19:23.206 [warning] Description: 'Server authenticity is not verified since certificate path validation is not enabled'
plausible_1  |      Reason: 'The option {verify, verify_peer} and one of the options \'cacertfile\' or \'cacerts\' are required to enable this.'
plausible_1  | 
plausible_1  | 08:19:23.310 [error] Task #PID<0.3645.0> started from #PID<0.3644.0> terminating
plausible_1  | ** (ArgumentError) raise/1 and reraise/2 expect a module name, string or exception as the first argument, got: {:network_failure, 'smtp.mailgun.org', {:error, :closed}}
plausible_1  |     (bamboo 2.2.0) lib/bamboo/strategies/task_supervisor_strategy.ex:25: anonymous fn/3 in Bamboo.TaskSupervisorStrategy.deliver_later/3
plausible_1  |     (elixir 1.14.3) lib/task/supervised.ex:89: Task.Supervised.invoke_mfa/2
plausible_1  |     (stdlib 4.2) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
plausible_1  | Function: #Function<0.115084270/0 in Bamboo.TaskSupervisorStrategy.deliver_later/3>
plausible_1  |     Args: []
plausible_1  | 08:19:57.497 request_id=F4MtNqD3wDjSmMYAAA7B [warning] Description: 'Server authenticity is not verified since certificate path validation is not enabled'
plausible_1  |      Reason: 'The option {verify, verify_peer} and one of the options \'cacertfile\' or \'cacerts\' are required to enable this.'
plausible_1  | 
plausible_1  | 08:19:57.600 request_id=F4MtNqD3wDjSmMYAAA7B [warning] Description: 'Server authenticity is not verified since certificate path validation is not enabled'
plausible_1  |      Reason: 'The option {verify, verify_peer} and one of the options \'cacertfile\' or \'cacerts\' are required to enable this.'
plausible_1  | 
plausible_1  | 08:19:57.704 request_id=F4MtNqD3wDjSmMYAAA7B [warning] Description: 'Server authenticity is not verified since certificate path validation is not enabled'
plausible_1  |      Reason: 'The option {verify, verify_peer} and one of the options \'cacertfile\' or \'cacerts\' are required to enable this.'
plausible_1  | 
plausible_1  | 08:19:57.807 request_id=F4MtNqD3wDjSmMYAAA7B [warning] Description: 'Server authenticity is not verified since certificate path validation is not enabled'
plausible_1  |      Reason: 'The option {verify, verify_peer} and one of the options \'cacertfile\' or \'cacerts\' are required to enable this.'
plausible_1  | 
plausible_1  | 08:19:57.909 request_id=F4MtNqD3wDjSmMYAAA7B [warning] Description: 'Server authenticity is not verified since certificate path validation is not enabled'
plausible_1  |      Reason: 'The option {verify, verify_peer} and one of the options \'cacertfile\' or \'cacerts\' are required to enable this.'
plausible_1  | 
plausible_1  | 08:19:58.013 request_id=F4MtNqD3wDjSmMYAAA7B [warning] Description: 'Server authenticity is not verified since certificate path validation is not enabled'
plausible_1  |      Reason: 'The option {verify, verify_peer} and one of the options \'cacertfile\' or \'cacerts\' are required to enable this.'
plausible_1  | 
```

**Reply by ruslandoga** [reply]
Ok, this is progress :) This error provides a clear path forward, we either need to add cacerts to verify `smtp.mailgun.org` or disable peer verification on ssl connection. The only problem is that it is not possible to do with just env vars and would require a PR + patch release.

**Reply by ruslandoga** [reply]
The PR would be similar to https://github.com/plausible/analytics/pull/3254 except for we can probably drop custom cacerts option and just hardcode `CAStore.file_path`

**Reply by ruslandoga** [reply]
I'll try to make these changes tomorrow and release a patch build on Monday.

**Reply by ruslandoga** [reply]
In the meantime you can probably use HTTP API endpoint to send emails through Mailgun. See https://plausible.io/docs/self-hosting-configuration for all the `MAILGUN_*` options, and note that you'd need to use `MAILER_ADAPTER=Bamboo.MailgunAdapter`

**Reply by gatodelcaribe** [reply]
Also tested deleting others lines on env and using it on docker-compose-yml.

code on docker compose:
```elixir
services:
  mail:
    image: bytemark/smtp
    restart: always
    environment:
      MAILER_ADAPTER: Bamboo.MailgunAdapter
      RELAY_HOST: smtp.mailgun.org
      MAILGUN_API_KEY: the mailgun api key
      RELAY_PORT: 465
      RELAY_USERNAME: analytics@email.gatodelcaribe.com
      RELAY_PASSWORD: password that mailgun provide me
``` 

Logs mail: 
```elixir
root@Plausible-analitycs:/opt/plausible# docker compose logs mail
mail_1  |     1 LOG: MAIN
mail_1  |     1   exim 4.89 daemon started: pid=1, -q30m, listening for SMTP on port 25 (IPv6 and IPv4)
mail_1  |   240 LOG: MAIN
mail_1  |   240   <= hello@plausible.local H=plausible_plausible_1.plausible_default (plausible.gatodelcaribe.com) [172.18.0.5] P=esmtp S=1826
mail_1  |   240 LOG: smtp_connection MAIN
mail_1  |   240   SMTP connection from plausible_plausible_1.plausible_default (plausible.gatodelcaribe.com) [172.18.0.5] closed by QUIT
mail_1  |   241 Exim version 4.89 uid=101 gid=101 pid=241 D=40001
mail_1  | Berkeley DB: Berkeley DB 5.3.28: (September  9, 2013)
mail_1  | Support for: crypteq iconv() IPv6 GnuTLS move_frozen_messages DKIM DNSSEC Event OCSP PRDR SOCKS TCP_Fast_Open
mail_1  | Lookups (built-in): lsearch wildlsearch nwildlsearch iplsearch cdb dbm dbmjz dbmnz dnsdb dsearch nis nis0 passwd
mail_1  | Authenticators: cram_md5 plaintext
mail_1  | Routers: accept dnslookup ipliteral manualroute queryprogram redirect
mail_1  | Transports: appendfile/maildir/mailstore autoreply lmtp pipe smtp
mail_1  | Fixed never_users: 0
mail_1  | Configure owner: 0:0
mail_1  | Size of off_t: 8
mail_1  |   241 delivering 1qetQp-00003s-Qj
mail_1  |   241 R: smarthost for kelbingato@gmail.com
mail_1  |   242 T: remote_smtp_smarthost for kelbingato@gmail.com
mail_1  |   242 Transport port=25 replaced by host-specific port=465
mail_1  |   242 Connecting to smtp.mailgun.org [34.160.157.95]:465 ... connected
```

**Reply by ruslandoga** [reply]
You don't need `mail` container if you are _**not**_ sending mail from the host. `MAILER_ADAPTER` and `MAILGUN_*` env vars would be set on `plausible` (which by default uses `plausible-conf.env` envfile) :)

**Reply by gatodelcaribe** [reply]
@ruslandoga I changed the port from "465" to "587" and now all emails working.

I saw that into the SMTP mailgun website it say: "Our servers listen on ports 25, 587, and 465 (SSL/TLS)"

**Reply by ruslandoga** [reply]
> This error provides a clear path forward, we either need to add cacerts to verify smtp.mailgun.org or disable peer verification on ssl connection. The only problem is that is not possible to do with just env vars and would require a PR + patch release.

Seems like [the adapter we use for SMTP](https://github.com/fewlinesco/bamboo_smtp) doesn't allow passing a custom hostname match function, so we wouldn't be able to properly configure it.

There are some unmerged PRs that make it possible:
- https://github.com/fewlinesco/bamboo_smtp/pull/223
- https://github.com/fewlinesco/bamboo_smtp/pull/210
- https://github.com/fewlinesco/bamboo_smtp/pull/214

`verify: :verify_none` is still an option but probably a bad one.
